src/lib/utils/error-logger.js

```diff
@@ -0,0 +1,40 @@
+'use strict';
+
+const logger = {};
+
+const getLogData = (err) => {
+    let logData;
+    if (err.stack) {
+        logData = err.stack;
+    } else if (err.message) {
+        logData = err.message;
+    } else if (err.toString instanceof Function) {
+        logData = err.toString();
+    } else {
+        logData = err;
+    }
+
+    return logData;
+};
+
+module.exports.register = (server, options, next) => {
+
+    //logs error
+    logger.logError = (tags,err) => {
+        server.log(['error'].concat(tags), getLogData(err));
+    };
+
+    //logs and throws error
+    logger.logAndThrowError = (tags,err) => {
```

Comment:
Your points are well taken. I agree that we can dispose with the `logAndThrow` method, and that should encourage better logging practices. Perhaps we could even make a general rule like 'all error logging should be confined to the route handler function'. 

With that out of the way, I'd like to refocus on your comment about server application error:
> In your provided example, providing a 401 is not a server application error--it's a client error. The server is not in an erroring state. It seems at face value to be dangerous to declare the server application in an error state when it is indeed working as expected.

 I think that some of our different viewpoints on the issue come from what we believe an `Error` object represents. I do not view Error objects strictly as indications of internal server errors (or server application error): They are indicative of any type of illegal state that the application finds itself in.

The nodeapi application uses Error objects to transmit information to the highest levels of the stack (to the Hapi.js internals that invoke the route handler, and reply to the client). As far as I know, Hapi endorses and encourages this behavior, as evidenced by their use of Boom: a library for decorating Error objects with information that will be useful to the client.

Similarly, promise rejections can be used to transmit more than just application errors: they can be used for flow control. Rejecting a promise via `Promise.reject()` or via `throw new Error()` allow us to declare the remaining steps of a promise chain as unnecessary. 

Just wanted to get that out there: does it make sense?